### PR TITLE
Expose CredentialProvider

### LINKED ISSUE #4163: Extensible ObjectStore Authentication
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

All the cloud providers provide CLIs for interacting with their service offering, a number of these come with fairly convoluted configuration setups for sourcing credentials. By and large these do not follow any widely adopted standard, and are inordinately complex.

**Describe the solution you'd like**

Rather than implementing all the various different authentication options, a truly herculean task, we could provide an option to shell out the CLI to obtain the credentials. @roeap added this for Azure in https://github.com/apache/arrow-rs/pull/3698.

Both AWS and GCP support this with

```
aws sts get-session-token
gcloud auth print-access-token
```

This would provide an escape hatch for CLI tools, without requiring a large amount of custom tooling. It would also provide complete coverage of all the options, including functionality that is not supported yet in any Rust library (e.g. GCP workload federation). Finally it would allow for workloads to choose to maximise compatibility by explicitly opting into using CLI-generated credentials #4137

**Describe alternatives you've considered**

**Additional context**

https://github.com/apache/arrow-rs/pull/2891 added experimental support for AWS_PROFILE

https://github.com/apache/arrow-rs/issues/3797 tracks adding support for more complex GCP authentication

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4163

# Rationale for this change
 

This exposes the CredentialProvider abstraction added in #4225. This allows users to provide custom mechanisms for sourcing credentials (#4163) and opening the door to exposing the authorisation logic (#4223)

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
cc @roeap

File: object_store/src/aws/mod.rs [hunk 1]
```diff
@@ -1002,76 +1017,74 @@ impl AmazonS3Builder {
...
-                    let token = WebIdentityProvider {
-                        token_path,
-                        session_name,
-                        role_arn,
-                        endpoint,
-                    };
-
-                    Arc::new(TokenCredentialProvider::new(
+        let credentials = if let Some(credentials) = self.credentials {
+            credentials
```
Comment (alamb) [inline object_store/src/aws/mod.rs:1021]: I think it  is worth considering  a `warn!` here if `access_key_id` or `secret_access_key` is set that they are being ignored in favor of the other credentials. I know the docs say this is the behavior but I think as a user I would find it somewhat confusing if the access_key_id got silently ignored -- I would rather the library loudly complained so I can fix it

Maybe in that vein, the builder should return an error if conflicting credentials are supplied
 🤔

File: object_store/src/azure/mod.rs [hunk 2]
```diff
@@ -937,7 +954,9 @@ impl MicrosoftAzureBuilder {
             let url = Url::parse(&account_url)
                 .context(UnableToParseUrlSnafu { url: account_url })?;
 
-            let credential = if let Some(bearer_token) = self.bearer_token {
+            let credential = if let Some(credential) = self.credentials {
```
Comment (alamb) [inline object_store/src/azure/mod.rs:957]: same comment here about erroring / warning

**Review by alamb — APPROVED** [review]
Makes sense to me. Thanks @tustvold

File: object_store/src/aws/mod.rs (hunk 1 above)
Comment (tustvold) [reply object_store/src/aws/mod.rs:1021]: I'm not sure about this as some systems may source config from the environment, which may then contain redundant credentials. I've filed https://github.com/apache/arrow-rs/issues/4240 to track this, and will come back to this after I've had some time to think on it, and others have had time to weigh in

**Pull request merged**